index.js

```diff
@@ -13,25 +14,45 @@ function HyperdriveSwarm (archive, opts) {
   this.uploading = !(opts.upload === false)
   this.downloading = !(opts.download === false)
   this.live = !(opts.live === false)
+  this.authorize = !(opts.authorize === false)
+
+  var isHyperdbInstance = !!(archive.get && archive.put && archive.replicate && archive.authorize)
+
+  if (isHyperdbInstance && !archive.local) {
+    throw new Error('hyperdiscovery swarm must be created after the local hyperdb instance is ready!')
+  }
 
   // Discovery Swarm Options
   opts = xtend({
     port: 3282,
-    id: archive.id,
+    id: isHyperdbInstance ? archive.local.id.toString('hex') : archive.id,
     hash: false,
     stream: function (peer) {
-      return archive.replicate({
+      return archive.replicate(xtend({
         live: self.live,
         upload: self.uploading,
         download: self.downloading
-      })
+      }, isHyperdbInstance ? {
+        userData: archive.local.key
+      } : {}))
     }
   }, opts)
 
   this.swarm = disc(swarmDefaults(opts))
   this.swarm.once('error', function () {
     self.swarm.listen(0)
   })
+
+  // Authorize peer connections if download is enabled...
+  this.downloading && this.swarm.on('connection', function (peer) {
+    if (peer.remoteUserData) {
+      archive.authorize(peer.remoteUserData, function () {
```

Comment:
this doesn't seem right? are you auto authenticating writers on every connect?

Reply:
Hmm, good point. What is the best approach for checking if a user is already authorized?

Reply:
I think you can just remove this line

Reply:
In order for outsiders to write I think it is need, but I may just have over-complicated it. I might have time this week, I'll take another look at it.

Reply:
Yes, but in some cases, the original creator might want the db to be read-only, or in case the key gets leaked, prevents spam from unwanted writers. So, writers are added explicitly.

Reply:
Someone can still extend this module by adding this code in the `.on('connection'` event.

Reply:
this .authorize() trick with userData is a fantastic hack that unlocks new potential

Reply:
@karissa Could I put it behind an option flag? So you have to opt-in to the auto-authorize? @mafintosh Is there any way to avoid authorizing the same people multiple times.

Reply:
@JimmyBoh I think auto-authorization is out of the scope of this module

Reply:
Sorry for the silence, I am working on removing the auto-authorization aspect now.